src/app/clusters/actions-server/actions-server.h

```diff
@@ -0,0 +1,341 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterface.h>
+#include <app/CommandHandlerInterface.h>
+#include <protocols/interaction_model/StatusCode.h>
+#include <vector>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace Actions {
+
+static constexpr size_t kActionNameMaxSize       = 128u;
+static constexpr size_t kEndpointListNameMaxSize = 128u;
+static constexpr size_t kEndpointListMaxSize     = 256u;
+
+class Delegate;
+
+struct ActionStructStorage : public Structs::ActionStruct::Type
+{
+    ActionStructStorage() : mActionName(mBuffer, sizeof(mBuffer)){};
+
+    ActionStructStorage(uint16_t action, CharSpan actionName, ActionTypeEnum actionType, uint16_t epListID,
+                        BitMask<CommandBits> commands, ActionStateEnum actionState) :
+        mActionName(mBuffer, sizeof(mBuffer))
+    {
+        Set(action, actionName, actionType, epListID, commands, actionState);
+    }
+
+    ActionStructStorage(const ActionStructStorage & action) : mActionName(mBuffer, sizeof(mBuffer)) { *this = action; }
+
+    ActionStructStorage & operator=(const ActionStructStorage & action)
+    {
+        Set(action.actionID, action.name, action.type, action.endpointListID, action.supportedCommands, action.state);
+        return *this;
+    }
+
+    void Set(uint16_t action, CharSpan actionName, ActionTypeEnum actionType, uint16_t epListID, BitMask<CommandBits> commands,
+             ActionStateEnum actionState)
+    {
+        actionID          = action;
+        type              = actionType;
+        endpointListID    = epListID;
+        supportedCommands = commands;
+        state             = actionState;
+        CopyCharSpanToMutableCharSpanWithTruncation(actionName, mActionName);
+        name = mActionName;
+    }
+
+private:
+    char mBuffer[kActionNameMaxSize];
+    MutableCharSpan mActionName;
+};
+
+struct EndpointListStorage : public Structs::EndpointListStruct::Type
+{
+    EndpointListStorage() : mEpListName(mBuffer, sizeof(mBuffer)){};
+
+    EndpointListStorage(uint16_t epListId, CharSpan epListName, EndpointListTypeEnum epListType,
+                        DataModel::List<const EndpointId> endpointList) :
```

Comment:
```suggestion
    EndpointListStorage(uint16_t epListId, const CharSpan & epListName, EndpointListTypeEnum epListType,
                        const DataModel::List<const EndpointId> & endpointList) :
```
and similar elsewhere here.